daffodil-test/src/test/resources/org/apache/daffodil/section07/discriminators/discriminator2.tdml

```diff
@@ -0,0 +1,150 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+<testSuite suiteName="NameDOB"
+           xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           xmlns:fn="http://www.w3.org/2005/xpath-functions"
+           xmlns:dfdl="http://www.ogf.org/dfdl/dfdl-1.0/"
+           xmlns:tdml="http://www.ibm.com/xmlns/dfdl/testData"
+           xmlns:ex="http://example.com"
+           xmlns:tns="http://example.com"
+           defaultRoundTrip="onePass">
+
+  <tdml:defineSchema name="s1">
+
+    <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />
+
+    <dfdl:format
+      ref="tns:GeneralFormat"
+      representation="text"
+      encoding="ASCII"
+      lengthKind="delimited"
+      separator=""
+      separatorPosition="infix"
+    />
+
+    <!--
+    Schema for simple CSV-like file containing 4 columns, the last of which is a date.
+
+    What makes this non-trivial is use of discriminator after the first column.
+    If we can parse a first column, then the remaining columns MUST be present, and an error
+    in parsing them should be fatal.
+
+    -->
+    <xs:element name="file" dfdl:initiator="last,first,middle,DOB%NL;%WSP*;">
+      <xs:complexType>
+        <xs:sequence dfdl:separator="%NL;" dfdl:separatorPosition="postfix">
+          <xs:element name="record" maxOccurs="unbounded">
```

Comment:
+1

I was curious to see if the CLI debugger would help shine some light on what's going on, and it actually was somewhat helpful now that it shows PoU's! Here's what's happening at a high level:

The first time a record is parsed we create a single PoU when the record element is created. The discriminator correctly resolves that PoU.

For the second and subsequent records, we actually create two PoU's when a record is parsed: one is for the record, one is for the the outer sequence (maybe related to separator suppression policy?). When the disciminator evaluates to true, it resolves only the record PoU but not the the sequence PoU. The existance of that outer PoU allows the parse to succeed and results in the left over data error.

So adding a second discriminator does allow this to work, though that probably shouldn't be required.

It also means that if there is an error parsing the first record, then you do get the expected error because there's only a single PoU for the first record. So whatever the solution, it's probably also worth adding a test where the error is on the first record, since there clearly is some difference in behavior between first and subsequent records right now.

Reply:
So you can't just put two discriminators, because only one is allowed at a given annotation point. 

The 2nd point of uncertainty is quite problematic. The separatorSuppression stuff really shouldn't add first class points of uncertainty. They're related in terms of the control flow of the parser, but likely should be some sort of kind that doesn't interact with discriminators.

This separatorSuppressionPolicy stuff is a very good reason to model things with terminators, not separators.

Reply:
You are correct that if the malformed date is first row, the failure is detected. 
I also made a variant that avoids separators, uses only terminators. 
Added tests that show this works, detects the malformed date in last row, as expected.

Reply:
I have thought for a long time that DFDL was missing an important dfdl:separatorSuppressionPolicy which should probably be the most common one which is "disabled", meaning that if the sequence contains 4 children there will be 3 separators (plus 1 if prefix or postfix). That is, empty-element suppression isn't even attempted. This is equivalent to adding a terminator to all the children of the sequence but the last (for infix).

Unfortunately separatorSuppressionPolicy 'never' doesn't mean this. It means that arrays first off must be bounded maxOccurs, and there will always be separators for that many occurrences.

Reply:
For DFDL v1.0 I think this means that one should model separation of sequences with terminators unless one wants some sort of suppression behavior.

Reply:
I'm not quite sure what is being said here, so I'll ask for clarification.
1. The goal of the schema is to fail the parse if an error occurs in the second, third, or fourth column of any line, yes?
2. An error should occur if a line has less than 4 columns, yes?
3. An error should occur if a line has more than 4 columns, yes?
4. An error should occur if the date in the 4th column can't be parsed, yes?
5. The terminator schema works, but the separator schema doesn't work as intended, yes?
6. Here's where I'm confused: is the problem a bug in Daffodil or a flaw in the DFDL 1.0 specification?
7. Should the separator schema work if the bug is fixed in Daffodil?
8. Or should the separator schema work if DFDL 1.0 allows the schema to say 
dfdl:separatorSuppressionPolicy="disabled" to avoid creating a second point of uncertainty?
9. Or does the problem go deeper; should the separator schema work if DFDL 1.0 allows separators to create second class points of uncertainty that don't interact with first class points of uncertainty created by discriminators?

Reply:
1 through 5 yes.
6 - the flaw in DFDL language is that this is so hard to figure out. One of the separatorSuppressionPolicies should correspond to "works exactly like terminators" if "postfix", and "infix" means no last terminator, and "prefix" means add an initiator before the first.
7 - yes, I believe so, but it will also allow some things to be considered well-formed that a user might think are malformed.
Such as allowing empty elements in places where the user wouldn't want them. 
8 - yes that policy would disallow things like excess empty elements such as sparse arrays allowed by separatorSuppressionPolicy trailingEmpty, anyEmpty, or trailingEmptyStrict
9 - This is a guess that having two kinds of uncertainty points might be a way to fix the problem and achieve (7).